src/jit/lir.cpp

```diff
@@ -1110,6 +1113,28 @@ bool LIR::Range::CheckLIR(Compiler* compiler, bool checkUnusedValues) const
     // Make sure the range itself is valid.
     assert(IsValid());
 
+    // Make sure the list is well-formed: no node appears more than once (no circularities). Insert each node in a set
+    // and assert we don't find the node again.
+    typedef SimplerHashTable<GenTreePtr, PtrKeyFuncs<GenTree>, bool, JitSimplerHashBehavior> NodeToBoolMap;
+    NodeToBoolMap* nodes = new (compiler->getAllocatorDebugOnly()) NodeToBoolMap(compiler->getAllocatorDebugOnly());
```

Comment:
Instead of using a map, this could use the LIR mark bit to track visited nodes. It's not critical, but it will save some  space (and potentially time, as the forward scan could be incorporated into the later checks). You could check that the bit is not set in the forward scan and that the bit is set in the reverse scan.